File: src/main/java/edu/practice/datastructures/hashtable/LinearProbingHashTable.java
package edu.practice.datastructures.hashtable;

public class LinearProbingHashTable<K, V> {

    static class Entry<K, V> {
        K key;
        V value;

        Entry(K key, V value) {
            this.key = key;
            this.value = value;
        }
    }

    private static final int INITIAL_CAPACITY = 16;
    private static final double LOAD_FACTOR_THRESHOLD = 0.7;

    private Entry<?, ?>[] table;

    private int size;

    public LinearProbingHashTable() {
        table = new Entry<?, ?>[INITIAL_CAPACITY];
        size = 0;
    }

    public int size() {
        return size;
    }

    public boolean isEmpty() {
        return size == 0;
    }

    private int hash(K key) {
        return (key != null) ? Math.abs(key.hashCode() % table.length) : -1;
    }

    @SuppressWarnings("unchecked")
    private void resize() {
        int newCapacity = table.length * 2;
        Entry<?, ?>[] oldTable = table;

        table = new Entry<?, ?>[newCapacity];
        size = 0;

        for (Entry<?, ?> entry : oldTable) {
            if (entry != null) {
                put((K) entry.key, (V) entry.value);
            }
        }
    }

    @SuppressWarnings("unchecked")
    public void put(K key, V value) {
        if ((double) size / table.length >= LOAD_FACTOR_THRESHOLD) {
            resize();
        }

        int index = hash(key);

        for (; table[index] != null; index = (index + 1) % table.length) {
            Entry<K, V> entry = (Entry<K, V>) table[index];
            if (entry.key.equals(key)) {
                entry.value = value; // Update existing key's value
                return;
            }
        }

        table[index] = new Entry<>(key, value);
        size++;
    }

    @SuppressWarnings("unchecked")
    public V get(K key) {
        int index = hash(key);

        for (; table[index] != null; index = (index + 1) % table.length) {
            Entry<K, V> entry = (Entry<K, V>) table[index];
            if (entry.key.equals(key)) {
                return entry.value;
            }
        }

        return null;
    }

    @SuppressWarnings("unchecked")
    public boolean remove(K key) {
        int index = hash(key);

        for (; table[index] != null; index = (index + 1) % table.length) {
            Entry<K, V> entry = (Entry<K, V>) table[index];
            if (entry.key.equals(key)) {
                entry.value = null; // Removal

                // Rehash subsequent keys
                index = (index + 1) % table.length;
                for (; table[index] != null; index = (index + 1) % table.length) {
                    Entry<K, V> rehashedEntry = (Entry<K, V>) table[index];

                    table[index] = null;
                    size--;

                    put(rehashedEntry.key, rehashedEntry.value); // Reinsert
                }

                return true;
            }
        }

        return false;
    }
}

File: src/main/kotlin/datastructures/hashtable/LinearProbingHashTable.kt
package edu.practice.datastructures.hashtable

import kotlin.math.abs

class LinearProbingHashTable<K, V>: HashTable<K, V> {

    companion object {
        private data class Entry<K, V>(val key: K, var value: V)

        private const val INITIAL_CAPACITY = 16
        private const val LOAD_FACTOR_THRESHOLD = 0.7
    }

    private var table: Array<Entry<K, V>?> = arrayOfNulls(INITIAL_CAPACITY)
    private var size = 0

    private fun hash(key: K): Int {
        return abs(key.hashCode() % table.size)
    }

    override fun put(key: K, value: V) {
        if (size.toDouble() / table.size >= LOAD_FACTOR_THRESHOLD) {
            resize()
        }

        var index = hash(key)
        while (table[index] != null) {
            val entry = table[index]
            if (entry?.key == key) {
                entry?.value = value // Update existing key's value
                return
            }

            index = (index + 1) % table.size
        }

        table[index] = Entry(key, value)
        size++
    }

    override fun get(key: K): V? {
        var index = hash(key)
        while (table[index] != null) {
            val entry = table[index]
            if (entry?.key == key) {
                return entry?.value
            }

            index = (index + 1) % table.size
        }

        return null
    }

    override fun remove(key: K): Boolean {
        var index = hash(key)
        while (table[index] != null) {
            val entry = table[index]
            if (entry?.key == key) {
                table[index] = null // Removal

                // Rehashing subsequent keys
                index = (index + 1) % table.size
                while (table[index] != null) {
                    val rehashedEntry = table[index]

                    table[index] = null
                    size--

                    rehashedEntry?.let { put(it.key, it.value) }
                    index = (index + 1) % table.size
                }

                return true
            }

            index = (index + 1) % table.size
        }

        return false
    }
    
    private fun resize() {
        val newCapacity = table.size * 2
        val oldTable: Array<Entry<K, V>?> = table

        table = arrayOfNulls(newCapacity)
        size = 0

        for (entry in oldTable) {
            entry?.let { put(it.key, it.value) }
        }
    }

    override fun size(): Int = size

    override fun isEmpty(): Boolean = (size == 0)
}